Merge feature/building-details

Ah-Ibrahim opened the pull request.

Ah-Ibrahim committed:
feat: add building repository

Ah-Ibrahim committed:
feat: add exceptions

Ah-Ibrahim committed:
feat: add exceptions handler

Ah-Ibrahim committed:
feat: add building service

Ah-Ibrahim committed:
feat: add UUID validation utility

Ah-Ibrahim committed:
chore(utils): renamed isValidUUID to isValidUUIDString for clarity

Ah-Ibrahim committed:
feat: add building response

Ah-Ibrahim committed:
feat: add building mapper along with floor mapper

Ah-Ibrahim committed:
feat: add building controller

Ah-Ibrahim committed:
chore(deps): update build.gradle.kts to handle JPA no-args constructor

Ah-Ibrahim committed:
fix: changed entity data class to class
Data classes are causing problems with JPA Hibernate

Ah-Ibrahim committed:
feat: add floor repository

Ah-Ibrahim committed:
feat: add floor service

Ah-Ibrahim committed:
feat: add floor dto and required location dto

Ah-Ibrahim committed:
feat: add floor controller

Ah-Ibrahim committed:
refactor: remove unused imports

Ah-Ibrahim committed:
refactor: remove redundant null checks

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/controller/v1/BuildingController.kt:
we need to add get all buildings end point

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/controller/v1/BuildingController.kt:
```diff
@@ -0,0 +1,28 @@
+
+@RestController
+@RequestMapping("/api/v1/building")
+class BuildingController(private val buildingService: BuildingService, private val buildingMapper: BuildingMapper) {
```
no need to make a mapper class, we can put the extension functions directly in mapper package

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/controller/v1/BuildingController.kt:
```diff
@@ -0,0 +1,28 @@
+class BuildingController(private val buildingService: BuildingService, private val buildingMapper: BuildingMapper) {
+	@GetMapping("/{id}")
+	fun getBuildingById(@PathVariable id: String): ResponseEntity<BuildingDetailsResponse> {
+		if (!UUIDUtils.isValidUUIDString(id)) throw InvalidIdFormException("Invalid building Id form")
```
it is a better if we depend on getUuidOrNull(), like this:
```
@GetMapping("/{id}")
    fun getBuildingById(@PathVariable id: String): ResponseEntity<BuildingDetailsResponse> {
        val uuid = getUuidOrNull(id) ?: throw InvalidIdFormException("Invalid building Id form")

        val building = buildingService.findBuildingById(uuid)
            ?: throw ResourceNotFoundException("Building with given Id is not found")

        return ResponseEntity.ok(buildingMapper.toBuildingDetailsResponse(building))
    }
```

```
or even this better:

fun getBuildingById(@PathVariable id: String): ResponseEntity<BuildingDetailsResponse> {
        return getUuidOrNull(id)?.let { uuid ->
            val building = buildingService.findBuildingById(uuid) ?: throw ResourceNotFoundException("Building with given Id is not found")
            ResponseEntity.ok(buildingMapper.toBuildingDetailsResponse(building))
        } ?: throw InvalidIdFormException("Invalid building Id form")
	}
```

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/controller/v1/FloorController.kt:
```diff
@@ -0,0 +1,28 @@
+class FloorController(private val floorService: FloorService, private val floorMapper: FloorMapper) {
+	@GetMapping("/{id}")
+	fun getFloorById(@PathVariable id: String): ResponseEntity<FloorDetailsResponse> {
+		if (!UUIDUtils.isValidUUIDString(id)) throw InvalidIdFormException("Invalid floor Id form")
```
same here, use getUuidOrNull better

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/controller/v1/FloorController.kt:
```diff
@@ -0,0 +1,28 @@
+
+@RestController
+@RequestMapping("/api/v1/floor")
+class FloorController(private val floorService: FloorService, private val floorMapper: FloorMapper) {
```
same for mapper, no need for extra class, just for simplicity

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/dto/LocationDto.kt:
no need for that class we could depend on LocationResponse(it hasn't been merged yet but was implemented in search location)

Ah-Ibrahim replied:
Ok, but I need it 😄

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/mapper/BuildingMapper.kt:
as I said no need for that class, we can make extension function directly like this:
fun Building.toBuildingDetailsResponse(): BuildingDetailsResponse{
    return BuildingDetailsResponse(
        id = this.id.toString(),
        name = this.name,
        imageUrl = this.imageUrl,
        floorsBasicDetails = this.floors.sortedBy { it.number }.map { floor ->
            floor.toFloorBasicDetails()
        }
    )
}

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/mapper/FloorMapper.kt:
```diff
@@ -0,0 +1,27 @@
+@Component
+class FloorMapper {
+	fun toBasicDetails(floor: Floor): FloorBasicDetails =
+		FloorBasicDetails(id = floor.id.toString(), number = floor.number)
```
avoid using compact function when the function is large, we use it only If the function is one line

yousef-osama11 commented on src/main/kotlin/com/indo/indo/api/mapper/FloorMapper.kt:
```diff
@@ -0,0 +1,27 @@
+			id = floor.toString(),
+			number = floor.number,
+			imageUrl = floor.imageUrl,
+			locations = floor.locations.sortedBy { it.name }.map {
```
it is better to give name to the lambda expression like this for better readability:
```
locations = floor.locations.sortedBy { it.name }.map { floor ->
				LocationByFloorDetails(
					id = floor.id.toString(),
					name = floor.name,
					aliasName = floor.aliasName,
				)
			}
```

yousef-osama11 commented on src/main/kotlin/com/indo/indo/entity/Building.kt:
```diff
@@ -9,28 +9,28 @@ import org.hibernate.annotations.UpdateTimestamp
-    @Id
-    @Column(columnDefinition = "uuid", updatable = false, nullable = false)
-    val id: UUID = UUID.randomUUID(),
+class Building(
```
why make it normal class ?!, it should be a data class

Ah-Ibrahim replied:
Data classes are not recommended to be used with Spring Data JPA, for several reasons:
- JPA requires classes to be open, while data classes are final by default
- JPA requires classes to have mutable properties, conflicting with data classes encouragement for immutability
- JPA needs no args constructor, while data classes enforces all constructor properties to be declared

sources:
- https://medium.com/%40goncharov.valentin/a-practical-point-of-view-on-kotlin-data-classes-and-jpa-hibernate-c69370b975e1
- https://garstecki.gitlab.io/articles/why-using-kotlin-data-classes-as-entities-is-a-bad-idea
- https://jpa-buddy.com/blog/best-practices-and-common-pitfalls

yousef-osama11 commented on src/main/kotlin/com/indo/indo/entity/Floor.kt:
```diff
@@ -8,29 +8,29 @@ import java.util.*
-    @Id
-    @Column(columnDefinition = "uuid", updatable = false, nullable = false)
-    val id: UUID = UUID.randomUUID(),
+class Floor(
```
same here, should be data class

yousef-osama11 commented on src/main/kotlin/com/indo/indo/entity/Location.kt:
```diff
@@ -8,30 +8,30 @@ import java.util.*
-    @Id
-    @Column(columnDefinition = "uuid", updatable = false, nullable = false)
-    val id: UUID = UUID.randomUUID(),
+class Location(
```
same here, data class should be used

yousef-osama11 commented on src/main/kotlin/com/indo/indo/util/UUIDUtils.kt:
there is already package called utills, so add all utils inside it

Ah-Ibrahim replied:
It is inside a package called util

yousef-osama11 commented on src/main/kotlin/com/indo/indo/util/UUIDUtils.kt:
```diff
@@ -0,0 +1,17 @@
+
+class UUIDUtils {
+	companion object Utils {
+		fun isValidUUIDString(uuid: String): Boolean {
```
we can use this function instead:
```
fun getUuidOrNull(id: String): UUID? =
    runCatching { UUID.fromString(id) }.getOrNull()
```

yousef-osama11 reviewed (commented):
great work @Ah-Ibrahim

Ah-Ibrahim committed:
fix: make name prop in LocationByFloorDetails dto nullable
Both name and alias name can be nullable, some locations have only name and other have only alias name.
This can cause a problem if both are null, but a database constraint has been added

Ah-Ibrahim committed:
refactor: substituted mapper classes with extension functions

Ah-Ibrahim committed:
fix: make invalid id form exception return 400 instead of 404
Invalid form requests should return 400 (Bad request) instead of 404.

Ah-Ibrahim committed:
refactor: change isValidUUIDString to getUUIDOrNull

Ah-Ibrahim committed:
refactor: removed unused imports

Ah-Ibrahim committed:
refactor: add parameter for lambda expressions for readability

Ah-Ibrahim committed:
feat: add get all buildings

yousef-osama11 commented on src/main/kotlin/com/indo/indo/util/UUIDUtils.kt:
```diff
@@ -0,0 +1,15 @@
+
+class UUIDUtils {
+	companion object Utils {
+		fun getUUIDOrNull(uuid: String): UUID? {
```
getUuidOrNull is better naming, also no need for class

Ah-Ibrahim committed:
refactor: rename getUuidOrNull for kotlin casing convention

Ah-Ibrahim committed:
refactor: rename DTOs
